cli/src/cast.rs

```diff
@@ -710,6 +728,51 @@ where
     })
 }
 
+fn det_base_in(value: &str, base_in: Option<String>) -> eyre::Result<u32> {
+    match base_in {
+        Some(base_in) => match base_in.as_str() {
+            "10" | "dec" => Ok(10),
+            "16" | "hex" => Ok(16),
+            _ => Err(eyre::eyre!("could not parse hex base in")),
+        },
+        None => {
+            if value.starts_with("0x") {
+                U256::from_str_radix(value, 16).expect("could not parse hex base in");
+
+                Ok(16)
+            } else {
+                match U256::from_str_radix(value, 10) {
+                    Ok(_) => {
+                        U256::from_str_radix(value, 16)
+                            .expect("could not parse base in, please prepend with 0x if hex.");
```

Comment:
i'm not sure about this logic - you're first checking if it can be decoded as base 10, if it can, then u check if it can be decoded as base 16?

Reply:
An hex number can always be parsed as a dec number but a dec cannot because it can contains letters.

So first check:
- Can it be parsed as a decimal number ?
    - Yes => Can it be parsed as hex ?
        - Yes => Then the base is ambiguous as the 0x is not here, it can either be hex or dec
        - No => It must be a dec
    - No => It may be an hex

Reply:
Yes, but this logic is more like

- Can it be parsed as decimal?
  - Yes: Can it be parsed as hex?
    - No: Fail
    - Yes: Assume decimal
  - No: It may be hex